README.md
# urlscraper

Download a list of URLs into a table.

# Development

First, set up your enviroment. It's just Docker.

Run `docker build .` to test that everything works.

Then do the Development Dance:

1. Write a failing test to `tests/`
2. Run `docker build .` to test it fails
3. Fix the code so all tests pass
4. Submit a pull request

# Deployment

1. Edit the version in `pyproject.toml`
2. Write an entry to `CHANGELOG.md`
3. Run `docker build .` one last time
4. `git push`


CHANGELOG.md
2021-01-28.01
-------------

* Move to a new GitHub repository.
* Support non-English error messages (table contents are still English)
* Download partial HTML and return it if request ultimately fails


tests/test_fetch.py
from typing import NamedTuple

import pandas as pd
from cjwmodule.testing.i18n import i18n_message
from freezegun import freeze_time
from pandas.testing import assert_frame_equal
from pytest_httpx import HTTPXMock

from urlscraper import fetch


def P(
    urlsource="list",
    urllist="",
    urlcol="",
    pagedurl="",
    addpagenumbers=False,
    startpage=0,
    endpage=9,
):
    return dict(
        urlsource=urlsource,
        urllist=urllist,
        urlcol=urlcol,
        pagedurl=pagedurl,
        addpagenumbers=addpagenumbers,
        startpage=startpage,
        endpage=endpage,
    )


class Settings(NamedTuple):
    SCRAPER_NUM_CONNECTIONS: int = 3
    SCRAPER_TIMEOUT: float = 1


async def _no_input_dataframe():
    return None


MOCK_DATETIME_S = "2021-01-28T18:56:12Z"


@freeze_time(MOCK_DATETIME_S)
def test_simple_urls(httpx_mock: HTTPXMock):
    httpx_mock.add_response(url="http://a.com/file", data=b"<div>all good</div>")
    httpx_mock.add_response(
        url="https://b.com/file2", status_code=404, data=b"not found"
    )
    httpx_mock.add_response(url="http://c.com/file/dir", data=b"<h1>What a page!</h1>")
    table, errors = fetch(
        P(urllist="http://a.com/file\nhttps://b.com/file2\nhttp://c.com/file/dir"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert_frame_equal(
        table,
        pd.DataFrame(
            {
                "url": [
                    "http://a.com/file",
                    "https://b.com/file2",
                    "http://c.com/file/dir",
                ],
                "date": [MOCK_DATETIME_S] * 3,
                "status": ["200 OK", "404 Not Found", "200 OK"],
                "html": ["<div>all good</div>", "not found", "<h1>What a page!</h1>"],
            }
        ),
    )


@freeze_time(MOCK_DATETIME_S)
def test_invalid_url():
    table, errors = fetch(
        P(urllist="just not a url\n/relative/url"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert_frame_equal(
        table,
        pd.DataFrame(
            {
                "url": ["http://just not a url", "http:///relative/url"],
                "date": [MOCK_DATETIME_S, MOCK_DATETIME_S],
                "status": ["Invalid URL", "Invalid URL"],
                "html": ["", ""],
            }
        ),
    )


def test_content_type_charset(httpx_mock: HTTPXMock):
    httpx_mock.add_response(
        data=b"caf\xe9", headers={"Content-Type": "text/html; charset=latin1"}
    )
    table, errors = fetch(
        P(urllist="http://hi"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert list(table["html"]) == ["café"]


def test_content_type_charset_invalid(httpx_mock: HTTPXMock):
    httpx_mock.add_response(
        data=b"caf\xe9", headers={"Content-Type": "text/html; charset=moo"}
    )
    table, errors = fetch(
        P(urllist="http://hi"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert list(table["html"]) == ["café"]  # fallback to latin1 because it won't fail


@freeze_time(MOCK_DATETIME_S)
def test_timeout_url(httpx_mock: HTTPXMock):
    # httpx_mock default is httpx.TimeoutException
    table, errors = fetch(
        P(urllist="https://example.org"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert_frame_equal(
        table,
        pd.DataFrame(
            {
                "url": ["https://example.org"],
                "date": [MOCK_DATETIME_S],
                "status": ["Timed out"],
                "html": [""],
            }
        ),
    )


# @freeze_time(MOCK_DATETIME_S)
# def test_no_connection_url(httpx_mock: HTTPXMock):
#     def raise_failure(request, ext):
#         raise httpx.ReadTimeout("failure", request=request)
#
#     httpx_mock.add_response(raise_failure, url="https://example.org")
#     table, errors = fetch(
#         P(urllist="https://example.org"),
#         get_input_dataframe=_no_input_dataframe,
#         settings=Settings(),
#     )
#     assert errors == []
#     assert_frame_equal(
#         table,
#         pd.DataFrame(
#             {
#                 "url": ["https://example.org"],
#                 "date": [MOCK_DATETIME_S],
#                 "status": ["failure"],
#                 "html": [""],
#             }
#         ),
#     )


def test_http_not_success(httpx_mock: HTTPXMock):
    httpx_mock.add_response(data=b"hi", status_code=429)
    table, errors = fetch(
        P(urllist="https://example.org"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert list(table["status"]) == ["429 Too Many Requests"]
    assert list(table["html"]) == ["hi"]


@freeze_time(MOCK_DATETIME_S)
def test_many_urls(httpx_mock: HTTPXMock):
    httpx_mock.add_response(url="http://1.com", data=b"1")
    httpx_mock.add_response(url="http://2.com", data=b"2")
    httpx_mock.add_response(url="http://3.com", data=b"3")
    httpx_mock.add_response(url="http://4.com", data=b"4")
    httpx_mock.add_response(url="http://5.com", data=b"5")
    url_list = ["http://%d.com" % i for i in range(1, 6)]
    table, errors = fetch(
        P(urllist="\n".join(url_list)),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(SCRAPER_NUM_CONNECTIONS=2),
    )
    assert errors == []
    assert_frame_equal(
        table,
        pd.DataFrame(
            {
                "url": url_list,
                "date": [MOCK_DATETIME_S] * 5,
                "status": ["200 OK"] * 5,
                "html": list("12345"),
            }
        ),
    )


def test_module_initial_nop():
    result = fetch(
        P(urlsource="list", urllist=""),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert result is None


def test_module_nop_with_initial_col_selection():
    async def get_input_dataframe():
        return pd.DataFrame({"A": [1]})

    result = fetch(
        P(urlsource="column", urlcol=""),
        get_input_dataframe=get_input_dataframe,
        settings=Settings(),
    )
    assert result is None


def test_urlcol(httpx_mock: HTTPXMock):
    async def get_input_dataframe():
        return pd.DataFrame({"A": ["http://a", "http://b"]})

    table, errors = fetch(
        P(urlsource="column", urlcol="A"),
        get_input_dataframe=get_input_dataframe,
        settings=Settings(),
    )
    assert errors == []
    assert list(table["url"]) == ["http://a", "http://b"]


def test_urlcol_truncate(httpx_mock: HTTPXMock):
    async def get_input_dataframe():
        return pd.DataFrame({"A": ["http://%i" % i for i in range(1, 12)]})

    table, errors = fetch(
        P(urlsource="column", urlcol="A"),
        get_input_dataframe=get_input_dataframe,
        settings=Settings(),
    )
    assert errors == [i18n_message("params.limitedNUrls", {"nUrls": 10})]
    assert list(table["url"]) == ["http://%i" % i for i in range(1, 11)]


def test_urlcol_missing_input_table(httpx_mock: HTTPXMock):
    result = fetch(
        P(urlsource="column", urlcol="A"),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert result is None


def test_scrape_list_truncate(httpx_mock: HTTPXMock):
    url_list = ["http://%d.com" % i for i in range(1, 12)]

    for url in url_list[:10]:
        httpx_mock.add_response(url=url, data=b"hi")

    table, errors = fetch(
        P(urllist="\n".join(url_list)),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )
    assert errors == [i18n_message("params.limitedNUrls", {"nUrls": 10})]
    assert_frame_equal(table[["url"]], pd.DataFrame({"url": url_list[:10]}))


def test_scrape_paged(httpx_mock: HTTPXMock):
    for i in range(1, 4):
        httpx_mock.add_response(url="http://foo?p=%d" % i, data=b"hi")

    table, errors = fetch(
        P(
            urlsource="paged",
            pagedurl="http://foo?p=",
            startpage=1,
            endpage=3,
            addpagenumbers=True,
        ),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )

    assert errors == []
    assert_frame_equal(
        table[["url"]],
        pd.DataFrame({"url": ["http://foo?p=1", "http://foo?p=2", "http://foo?p=3"]}),
    )


def test_scrape_paged_truncate(httpx_mock: HTTPXMock):
    for i in range(1, 11):
        httpx_mock.add_response(url="http://foo?p=%d" % i, data=b"hi")

    table, errors = fetch(
        P(
            urlsource="paged",
            pagedurl="http://foo?p=",
            startpage=1,
            endpage=23,
            addpagenumbers=True,
        ),
        get_input_dataframe=_no_input_dataframe,
        settings=Settings(),
    )

    assert errors == [i18n_message("params.limitedNUrls", {"nUrls": 10})]
    assert_frame_equal(
        table[["url"]],
        pd.DataFrame({"url": ["http://foo?p=%d" % i for i in range(1, 11)]}),
    )


tests/test_migrate_params.py
from urlscraper import migrate_params


def test_v1():
    assert migrate_params(
        {"urlsource": 0, "urlcol": "A", "urllist": "http://example.org\n"}
    ) == {
        "urlsource": "list",
        "urlcol": "A",
        "urllist": "http://example.org\n",
        "pagedurl": "",
        "addpagenumbers": False,
        "startpage": 0,
        "endpage": 9,
    }


def test_v2():
    assert migrate_params(
        {"urlsource": "list", "urlcol": "A", "urllist": "http://example.org\n"}
    ) == {
        "urlsource": "list",
        "urlcol": "A",
        "urllist": "http://example.org\n",
        "pagedurl": "",
        "addpagenumbers": False,
        "startpage": 0,
        "endpage": 9,
    }


def test_v3():
    # test that moving form v2 to v3 sets addpagenumbers to True, as that was what v2 did
    assert migrate_params(
        {
            "urlsource": "paged",
            "urlcol": "A",
            "urllist": "",
            "pagedurl": "http://example.org/foo?page=",
            "startpage": 1,
            "endpage": 4,
        }
    ) == {
        "urlsource": "paged",
        "urlcol": "A",
        "urllist": "",
        "pagedurl": "http://example.org/foo?page=",
        "addpagenumbers": True,
        "startpage": 1,
        "endpage": 4,
    }


urlscraper.py
import asyncio
import datetime
import io
import re
from typing import List, Protocol, Tuple

import pandas as pd
import rfc3987
from cjwmodule.http import HttpError
from cjwmodule.http.client import download
from cjwmodule.http.httpfile import extract_first_header
from cjwmodule.i18n import trans

MaxNUrls = 10


class Settings(Protocol):
    SCRAPER_NUM_CONNECTIONS: int
    """Maximum number of concurrent fetches."""

    SCRAPER_TIMEOUT: float
    """Maximum number of seconds before abandoning HTTP request."""


def utcnow():
    """
    Return datetime.datetime.utcnow().

    It's a separate function so we can mock it in unit tests.
    """
    return datetime.datetime.utcnow()


def _as_text(buf: bytes, headers: List[Tuple[str, str]]) -> str:
    """Guess encoding and return buf as text.

    Never raise.
    """
    if isinstance(headers, list):
        content_type = extract_first_header(headers, "Content-Type")
    else:
        # httpx Response headers
        content_type = headers.get("Content-Type")
    if content_type and "charset=" in content_type:
        charset = content_type.split("charset=")[1]
    else:
        charset = "utf-8"  # guessing
    try:
        return buf.decode(encoding=charset, errors="replace")
    except LookupError:  # invalid encoding
        return buf.decode("latin1")  # never fails


async def async_get_url(row, url, *, settings: Settings):
    """Return a Future (row, status, text).

    The Future will resolve within settings.SCRAPER_TIMEOUT seconds. `status`
    may be one of '200 OK', '404 Not Found', 'Timed out', 'Invalid URL',
    "Can't connect: [err]", 'Unknown error: [err]'.
    """
    buf = io.BytesIO()

    def ret(status_text: str, headers: List[Tuple[str, str]] = []):
        return row, status_text, _as_text(buf.getvalue(), headers)

    try:
        parts = rfc3987.parse(url, "absolute_URI")  # raise ValueError
        if parts["scheme"] not in {"http", "https"}:
            raise ValueError("Unsupported scheme")
        if not parts["authority"]:
            raise ValueError(
                "No authority, probably because we prepended 'http:' to any old string"
            )
    except ValueError:
        return ret("Invalid URL")

    try:
        downloaded = await download(url, buf, total_timeout=settings.SCRAPER_TIMEOUT)
        return ret(
            "%d %s" % (downloaded.status_code, downloaded.reason_phrase),
            downloaded.headers,
        )
    except HttpError.InvalidUrl:
        return ret("Invalid URL")
    except HttpError.Timeout:
        return ret("Timed out")
    except HttpError.TooManyRedirects:
        return ret("Too many redirects")
    except HttpError.NotSuccess as err:
        return ret(
            "%d %s" % (err.response.status_code, err.response.reason_phrase),
            err.response.headers,
        )
    except HttpError.Generic as err:
        return ret(str(err))


# Asynchronously scrape many urls, and store the results in the table
async def scrape_urls(urls, result_table, settings):
    next_queued_row = 0  # index into urls
    fetching = set()  # {Future<response>}

    max_fetchers = settings.SCRAPER_NUM_CONNECTIONS

    while next_queued_row < len(urls) or fetching:
        # start tasks until we max out connections, or run out of urls
        while next_queued_row < len(urls) and len(fetching) < max_fetchers:
            row = next_queued_row
            url = urls[row].strip()
            fetching.add(async_get_url(row, url, settings=settings))

            next_queued_row += 1

        assert fetching

        # finish one or more tasks, then loop
        done, pending = await asyncio.wait(
            fetching, return_when=asyncio.FIRST_COMPLETED
        )

        for task in done:
            row, status, text = await task
            result_table.loc[row, "status"] = status
            result_table.loc[row, "html"] = text

        fetching = pending  # delete done tasks


def render(table, params, *, fetch_result):
    if fetch_result is None:
        return table

    else:
        return fetch_result


def _truncate_error():
    return trans(
        "params.limitedNUrls",
        "We limited your scrape to {nUrls} URLs",
        {"nUrls": MaxNUrls},
    )


def fetch(params, *, get_input_dataframe, settings):
    urls = []
    urlsource = params["urlsource"]
    errors = []

    if urlsource == "list":
        urllist_text: str = params["urllist"]
        urllist_raw = urllist_text.split("\n")
        for url in urllist_raw:
            s_url = url.strip()
            if len(s_url) == 0:
                continue
            # Fix in case user adds an URL without http(s) prefix
            if not re.match("^https?://.*", s_url):
                urls.append("http://{}".format(s_url))
            else:
                urls.append(s_url)
        if not urls:
            return None
        if len(urls) > MaxNUrls:
            urls = urls[:MaxNUrls]
            errors.append(_truncate_error())
    elif urlsource == "column":
        # We won't execute here -- there's no need: the user clicked a
        # button so should be pretty clear on what the input is.
        prev_table = asyncio.run(get_input_dataframe())

        if prev_table is None or params["urlcol"] not in prev_table.columns:
            return None

        # get our list of URLs from a column in the input table
        urlcol: str = params["urlcol"]
        urls = prev_table[urlcol].tolist()
        if len(urls) > MaxNUrls:
            urls = urls[:MaxNUrls]
            errors.append(_truncate_error())
    elif urlsource == "paged":
        # Count through a list of page numbers, appending each to the URL
        pagedurl: str = params["pagedurl"]
        if not pagedurl:
            return None

        # Fix in case user adds an URL without http(s) prefix
        if not re.match("^https?://.*", pagedurl):
            pagedurl = "http://" + pagedurl

        begin = params["startpage"]
        end = params["endpage"] + 1
        if end - begin > MaxNUrls:
            end = begin + MaxNUrls
            errors.append(_truncate_error())

        # Generate multiple urls by adding page numbers, if user says so
        if params["addpagenumbers"]:
            # limit the number of pages we can scrape with this method
            urls = [pagedurl + str(num) for num in range(begin, end)]
        else:
            urls = [pagedurl]
    else:
        raise RuntimeError("Invalid urlsource")

    table = pd.DataFrame(
        {
            "url": urls,
            # TODO use response date, not current date
            # TODO migrate to use timestamp type, not text (will affect
            # existing users)
            "date": [utcnow().isoformat(timespec="seconds") + "Z"] * len(urls),
            "status": [""] * len(urls),
            "html": [""] * len(urls),
        }
    )

    asyncio.run(scrape_urls(urls, table, settings=settings))
    return table, errors


def _migrate_params_v0_to_v1(params):
    """
    v0: urlsource was 0 ("List") or 1 ("Input column")

    v1: urlsource is "list" or "column".
    """
    return {**params, "urlsource": ["list", "column"][params["urlsource"]]}


def _migrate_params_v1_to_v2(params):
    """
    v2 adds "paged" option to urlsource menu and related parameters
    """
    return {
        **params,
        "pagedurl": "",
        "addpagenumbers": False,  # defaults, from json file
        "startpage": 0,
        "endpage": 9,
    }


def _migrate_params_v2_to_v3(params):
    """
    v3 adds "addpagenumbers" checkbox
    """
    return {**params, "addpagenumbers": True}  # match v2 behavior


def migrate_params(params):
    if isinstance(params["urlsource"], int):
        params = _migrate_params_v0_to_v1(params)
    if "pagedurl" not in params:
        params = _migrate_params_v1_to_v2(params)
    if "addpagenumbers" not in params:
        params = _migrate_params_v2_to_v3(params)
    return params
